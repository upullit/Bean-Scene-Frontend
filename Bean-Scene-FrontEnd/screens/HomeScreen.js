import React, { useEffect, useState } from 'react';
import { View, Text, TouchableOpacity, StyleSheet, Modal, ScrollView } from 'react-native';
import * as ScreenOrientation from 'expo-screen-orientation';
import { useAuth } from '../context/authContext';
import CustomModal from '../CustomModal';

const HomeScreen = ({ navigation }) => {
    const { user, logout } = useAuth();
    const [logoutModalVisible, setLogoutModalVisible] = useState(false);

    // Lock the app orientation
    useEffect(() => {
        const lockOrientation = async () => {
            try {
                await ScreenOrientation.lockAsync(ScreenOrientation.OrientationLock.LANDSCAPE);
            } catch (err) {
                console.warn("Orientation lock not supported:", err.message);
            }
        };
        lockOrientation();

        return () => {
            ScreenOrientation.unlockAsync(); // Reset orientation on unmount
        };
    }, []);

    useEffect(() => {
        console.log('User object in HomeScreen:', user); // Check what is being passed
    }, [user]);


    const handleLogout = async () => {
        await logout(); // Clear the user state
        setLogoutModalVisible(false); // Close the logout modal
    };

    return (
        <View>
            <View style={styles.container}>
                <Text style={styles.title}>Bean Scene Ordering App</Text>
                <Text style={styles.text}>
                    Welcome, {user ? `${user.name} Role: ${user.role}` : 'Guest'}
                </Text>

                {/* Role-based navigation buttons */}
                {user?.role === 'Server' && (
                    <>
                        <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('ServerOrder')}>
                            <Text style={styles.buttonText}>Server Ordering</Text>
                        </TouchableOpacity>
                        <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('Ticket')}>
                            <Text style={styles.buttonText}>Tickets</Text>
                        </TouchableOpacity>
                    </>
                )}

                {user?.role === 'Manager' && (
                    <>
                        <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('ServerOrder')}>
                            <Text style={styles.buttonText}>Server Ordering</Text>
                        </TouchableOpacity>
                        <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('Ticket')}>
                            <Text style={styles.buttonText}>Tickets</Text>
                        </TouchableOpacity>
                        <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('Admin')}>
                            <Text style={styles.buttonText}>Admin</Text>
                        </TouchableOpacity>
                        </>
                )}

                {user?.role === 'Chef' && (
                    <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('Kitchen')}>
                        <Text style={styles.buttonText}>Kitchen Ticket</Text>
                    </TouchableOpacity>
                )}

                {/* Accessible to all staff */}
                <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('CustomerMenu')}>
                    <Text style={styles.buttonText}>Customer Menu</Text>
                </TouchableOpacity>                
                {!user && (
                    <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('Login')}>
                        <Text style={styles.buttonText}>Login</Text>
                    </TouchableOpacity>
                )}
                {!user && (
                    <TouchableOpacity style={styles.button} onPress={() => navigation.navigate('RegisterStaff')}>
                        <Text style={styles.buttonText}>Register</Text>
                    </TouchableOpacity>
                )}
                {/* Logout button */}
                {user && (
                    <TouchableOpacity
                        style={styles.button}
                        onPress={() => setLogoutModalVisible(true)}
                    >
                        <Text style={styles.buttonText}>Logout</Text>
                    </TouchableOpacity>
                )}

                {/* Logout Confirmation Modal */}
                <Modal
                    transparent={true}
                    visible={logoutModalVisible}
                    animationType="slide"
                    onRequestClose={() => setLogoutModalVisible(false)}>
                    <View style={styles.overlay}>
                        <View style={styles.modalContainer}>
                            <Text style={styles.modalText}>Are you sure you want to log out?</Text>
                            <View style={styles.modalButtons}>
                                <TouchableOpacity
                                    style={[styles.modalButton, styles.cancelButton]}
                                    onPress={() => setLogoutModalVisible(false)}>
                                    <Text style={styles.buttonText}>Cancel</Text>
                                </TouchableOpacity>
                                <TouchableOpacity
                                    style={[styles.modalButton, styles.confirmButton]}
                                    onPress={handleLogout}>
                                    <Text style={styles.buttonText}>Log Out</Text>
                                </TouchableOpacity>
                            </View>
                        </View>
                    </View>
                </Modal>
            </View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        alignContent: 'center',
        justifyContent: 'center',
        padding: 20,
        marginTop: 20,
        marginLeft: 200,
        marginRight: 200,
        marginBottom: 200,
    },
    title: {
        fontSize: 32,
        marginTop: 10,
        fontWeight: 'bold',
        marginBottom: 50,
        textAlign: 'center',
    },
    button: {
        backgroundColor: '#76453B',
        padding: 15,
        marginVertical: 10,
        borderRadius: 5,
    },
    buttonText: {
        color: '#F8FAE5',
        fontSize: 20,
        textAlign: 'center',
        fontWeight: 'bold',
    },
    text: {
        fontSize: 20,
        textAlign: 'center',
        fontWeight: 'bold',
        color: '#251605',
    },
    overlay: {
        flex: 1,
        backgroundColor: "rgba(0, 0, 0, 0.5)",
        justifyContent: "center",
        alignItems: "center",
      },
      modalContainer: {
        width: 400,
        backgroundColor: '#F8FAE5',
        borderRadius: 10,
        padding: 20,
        alignItems: "center",
        elevation: 5,
      },
    modalText: {
        fontSize: 18,
        marginBottom: 10,
        textAlign: 'center',
    },
    modalButtons: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        width: '100%',
    },
    modalButton: {
        flex: 1,
        padding: 10,
        margin: 5,
        borderRadius: 5,
        alignItems: 'center',
    },
    cancelButton: {
        backgroundColor: '#d9534f',
    },
    confirmButton: {
        backgroundColor: '#5cb85c',
    },
});

export default HomeScreen;
