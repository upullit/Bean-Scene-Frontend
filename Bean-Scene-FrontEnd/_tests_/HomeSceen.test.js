import React from 'react';
import { Text, TouchableOpacity, View, StyleSheet } from 'react-native';
import renderer from 'react-test-renderer'; 
import HomeScreen from '../screens/HomeScreen'; // Adjust the import path as needed
import * as ScreenOrientation from 'expo-screen-orientation';

jest.mock('expo-screen-orientation', () => ({
  lockAsync: jest.fn(),
  unlockAsync: jest.fn(),
  OrientationLock: {
    LANDSCAPE: 'LANDSCAPE',
  },
}));

describe('HomeScreen', () => {
  const mockNavigation = {
    navigate: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly', () => {
    const tree = renderer.create(<HomeScreen navigation={mockNavigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('calls lockAsync on mount', () => {
    const component = renderer.create(<HomeScreen navigation={mockNavigation} />);
    // Ensure lockAsync is called with LANDSCAPE
    expect(ScreenOrientation.lockAsync).toHaveBeenCalledWith(ScreenOrientation.OrientationLock.LANDSCAPE);
  });

  test('navigates to the correct screens on button presses', () => {
    const component = renderer.create(<HomeScreen navigation={mockNavigation} />);
    const root = component.root;

    // Find all TouchableOpacity elements
    const buttons = root.findAllByType(TouchableOpacity);

    // Simulate button presses
    buttons[0].props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith('ServerOrder');

    buttons[1].props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Ticket');

    buttons[2].props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Kitchen');

    buttons[3].props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Admin');

    buttons[4].props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith('CustomerMenu');

    buttons[5].props.onPress();
    expect(mockNavigation.navigate).toHaveBeenCalledWith('CustomerOrder');
  });

  // New test for style assertions
  test('has correct styles applied to elements', () => {
    const component = renderer.create(<HomeScreen navigation={mockNavigation} />);
    const root = component.root;

    // Find elements to test their styles
    const texts = root.findAllByType(Text);
    const title = texts[0];
    expect(title.props.style).toEqual(expect.objectContaining({
      fontSize: 32,
      fontWeight: 'bold',
      textAlign: 'center',
      marginBottom: 50,
    }));
  
    // Check the general description text style (second Text element)
    const descriptionText = texts[1]; // Assuming this is the description "Hi, this is the Bean Scene menu ordering app"
    expect(descriptionText.props.style).toEqual(expect.objectContaining({
      fontSize: 20,
      fontWeight: 'bold',
      textAlign: 'center',
      color: '#251605',
    }));
  
    // Check the button text style (third Text element)
    const buttonText = texts[2]; // This should be the first button text, "Server Ordering"
    expect(buttonText.props.style).toEqual(expect.objectContaining({
      color: '#F8FAE5',
      fontSize: 20,
      textAlign: 'center',
      fontWeight: 'bold',
    }));
  
    // Check all button styles (TouchableOpacity elements)
    const buttons = root.findAllByType(TouchableOpacity);
    buttons.forEach((button) => {
      expect(button.props.style).toEqual(expect.objectContaining({
        backgroundColor: '#76453B',
        padding: 15,
        marginVertical: 10,
        borderRadius: 5,
      }));
    });
  })});