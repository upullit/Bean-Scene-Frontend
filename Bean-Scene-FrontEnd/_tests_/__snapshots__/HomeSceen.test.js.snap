// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly 1`] = `
<View>
  <View
    style={
      {
        "alignContent": "center",
        "justifyContent": "center",
        "marginBottom": 200,
        "marginLeft": 200,
        "marginRight": 200,
        "marginTop": 20,
        "padding": 20,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 32,
          "fontWeight": "bold",
          "marginBottom": 50,
          "marginTop": 10,
          "textAlign": "center",
        }
      }
    >
      Bean Scene Ordering App
    </Text>
    <Text
      style={
        {
          "color": "#251605",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Hi, this is the Bean Scene menu ordering app
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#76453B",
          "borderRadius": 5,
          "marginVertical": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          {
            "color": "#F8FAE5",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Server Ordering
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#76453B",
          "borderRadius": 5,
          "marginVertical": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          {
            "color": "#F8FAE5",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Tickets
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#76453B",
          "borderRadius": 5,
          "marginVertical": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          {
            "color": "#F8FAE5",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Kitchen Ticket
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#76453B",
          "borderRadius": 5,
          "marginVertical": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          {
            "color": "#F8FAE5",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Admin
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#76453B",
          "borderRadius": 5,
          "marginVertical": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          {
            "color": "#F8FAE5",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Customer Menu
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#76453B",
          "borderRadius": 5,
          "marginVertical": 10,
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          {
            "color": "#F8FAE5",
            "fontSize": 20,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Customer Order
      </Text>
    </View>
  </View>
</View>
`;
